store/src/directory.rs

```diff
@@ -256,30 +536,94 @@ where
+            Ok(file
+                .set_xattr(
+                    format_xattr(metadata_key.to_key()),
+                    &metadata_value.to_stored()?,
+                )
+                .map_err(|e| {
+                    StoreError::Unspecified(format!(
+                        "Error creating xattr on {}: {:?}",
+                        path.display(),
+                        e
+                    ))
+                })?)
+        } else {
+            let path = set_metadata_extension_to_path(&path)?;
+            let (file, previous) = match File::open(&path) {
+                Ok(f) => (f, true),
+                Err(e) if e.kind() == std::io::ErrorKind::NotFound => match File::create(&path) {
+                    Err(e) => {
+                        return Err(StoreError::Unspecified(format!(
+                            "Couldn't create file: {}",
+                            e
+                        )));
+                    }
+                    Ok(f) => (f, false),
+                },
+                Err(e) => {
+                    return Err(StoreError::Unspecified(format!(
+                        "Error opening file (r): {}",
+                        e,
+                    )));
+                }
+            };
+            let mut metadata: FdoMetadata;
+            if previous {
+                metadata = match serde_cbor::from_reader(&file) {
+                    Ok(data) => data,
+                    Err(e) => {
+                        return Err(StoreError::Unspecified(format!(
+                            "Error deserialising data: {}",
+                            e
+                        )))
+                    }
+                };
+                metadata.map.insert(
+                    metadata_key.to_key().to_string(),
+                    metadata_value.to_stored()?,
+                );
+            } else {
+                let mut hm = HashMap::new();
+                hm.insert(
+                    metadata_key.to_key().to_string(),
+                    metadata_value.to_stored()?,
+                );
+                metadata = FdoMetadata { map: hm };
             }
-            Ok(f) => f,
-        };
-
-        Ok(file
-            .set_xattr(
-                format_xattr(metadata_key.to_key()),
-                &metadata_value.to_stored()?,
-            )
-            .map_err(|e| {
-                StoreError::Unspecified(format!(
-                    "Error creating xattr on {}: {:?}",
-                    path.display(),
-                    e
-                ))
-            })?)
+            drop(file);
+            let file = match File::create(&path) {
+                Ok(f) => f,
+                Err(e) => {
+                    return Err(StoreError::Unspecified(format!(
+                        "Error opening file (w): {}",
+                        e
+                    )))
+                }
+            };
+            match serde_cbor::to_writer(&file, &metadata) {
```

Comment:
The spec uses cbor explicitly

Reply:
this is not anything spec related afaict, this is how we store metadata and storing it in cbor really makes things difficult to debug if not strictly necessary

Reply:
I simply used cbor because it is what we were using due to the standard everywhere else, so only due to consistency reasons. If we want the metadata files to be readable we could use json.